controller/volume_controller.go

```diff
@@ -3022,6 +3022,153 @@ func (vc *VolumeController) ResolveRefAndEnqueue(namespace string, ref *metav1.O
+func restoreVolumeRecurringJobLabel(v *longhorn.Volume, jobName string, job *longhorn.RecordVolumeRecurringJob) error {
+	// set Volume recurring jobs/groups
+	if job.FromJob {
+		labelKey := types.GetRecurringJobLabelKey(types.LonghornLabelRecurringJob, jobName)
+		// enable recurring jobs
+		v.Labels[labelKey] = types.LonghornLabelValueEnabled
+	}
+	for _, groupName := range job.FromGroup {
+		groupLabelKey := types.GetRecurringJobLabelKey(types.LonghornLabelRecurringJobGroup, groupName)
+		// enable recurring groups
+		v.Labels[groupLabelKey] = types.LonghornLabelValueEnabled
+	}
+
+	return nil
+}
+
+// createRecurringJobNotExist add recurring jobs and group back to a restoring volume and create recurring jobs if not exist
+func createRecurringJobNotExist(log *logrus.Entry, ds *datastore.DataStore, v *longhorn.Volume, jobName string, job *longhorn.RecordVolumeRecurringJob) (string, error) {
+	if existJob, err := ds.GetRecurringJob(jobName); err != nil {
+		if !apierrors.IsNotFound(err) {
+			log.WithError(err).Warnf("Failed to get the recurring job %v", jobName)
+			return "", err
+		}
+	} else if !reflect.DeepEqual(existJob.Spec, job.JobSpec) {
+		// create a new recurring job if job name is used and configuration is different.
+		newJobName, existJobName, err := generateRecurringJobName(log, ds, job)
+		if err != nil {
+			log.WithError(err).Warnf("Failed to create a new recurring job %v", jobName)
+			return "", err
+		}
+		if existJobName != "" {
+			return existJobName, nil
+		}
+		jobName = newJobName
+	} else {
+		return jobName, nil
+	}
+
+	// create the recurring job if not exist
+	log.Debugf("Creating the recurring job %v when restoring recurring jobs", jobName)
+	if _, err := ds.CreateRecurringJob(&longhorn.RecurringJob{ObjectMeta: metav1.ObjectMeta{Name: jobName}, Spec: job.JobSpec}); err != nil {
+		log.WithError(err).Warnf("Failed to create a recurring job %v", jobName)
+		return "", err
+	}
+
+	return jobName, nil
+}
+
+// restoreVolumeRecurringJobs create recurring jobs/groups from the backup volume when restoring from a backup, except for the DR volume
+func (vc *VolumeController) restoreVolumeRecurringJobs(v *longhorn.Volume) error {
+	log := getLoggerForVolume(vc.logger, v)
+	backupVolumeRecurringJobsInfo := make(map[string]longhorn.RecordVolumeRecurringJob)
+	bvName, isExist := v.Labels[types.LonghornLabelBackupVolume]
+	if !isExist {
+		log.Warn("Cannot find the backup volume label")
+		return nil
+	}
+	bv, err := vc.ds.GetBackupVolumeRO(bvName)
+	if err != nil {
+		log.WithError(err).WithField("backupvolume", bvName).Warn("Failed to get the backup volume info")
+		return err
+	}
+
+	if volumeRecurringJobInfoStr, exist := bv.Status.Labels[types.VolumeRecurringJobInfoLabel]; exist {
+		if err := json.Unmarshal([]byte(volumeRecurringJobInfoStr), &backupVolumeRecurringJobsInfo); err != nil {
+			log.WithError(err).WithField("backupvolume", bvName).Warn("Failed to unmarshal information of volume recurring jobs")
```

Comment:
NIT: This will fail the function. Which log-level should we use, warning or error?

Reply:
Modified. Thanks.